### PR TITLE
[Security]  Add security detection engine package

### PR DESCRIPTION
**Pull request by rw-access**
## What does this PR do?

Added an integration that contains `security_rule` assets to be used by the Detection Engine in Security. The core idea behind this effort is to use Fleet to update rules for the Detection Engine between releases. This way, users can get the latest released rules and package-relevant rules without needing to wait  for the next stack release.

See the "related issues" section or ping me for more context. I'll be happy to point to additional (internal) design docs and discussions.

## Checklist

- [X] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them. (have not tested the package using `elastc-package` yet)
- [ ] ~I have verified that all data streams collect metrics or logs.~
- [x] I have added an entry to my package's `changelog.yml` file.

## Author's Checklist

We've had a lot of conversations between teams, so I think we're in alignment. The name of the integration is TBD,  and the specifics to install the `security_rule` asset type are also undecided.

## How to test this PR locally

There is a chicken-and-egg issue here. Since it contains a brand new asset type, Kibana doesn't yet know how to install it. So the package will be published first as a pre-release and once this merges and is added to package-storage, then we can start developing the Kibana side to install the package.

There will be more iteration as we go between all of the repositories.

## Related issues

*https://github.com/elastic/package-spec/pull/142: Add `security_rule` asset type
* https://github.com/elastic/package-spec/pull/151: Add `NOTICE.txt` file
* https://github.com/elastic/package-storage/pull/843: Original package for this to power 
* https://github.com/elastic/kibana/pull/91949: Now closed, initially pulled in https://github.com/elastic/package-storage/pull/843, but our design has changed so that EPM will publish the rules as saved objects to an intermediate staging area

## Screenshots

None yet

### CODE REVIEW

File: packages/security_detection_engine/changelog.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# newer versions go on top
+- version: "0.0.1" # 0.0.1-dev.1 fails validation
```
Comment (mtojek) [inline packages/security_detection_engine/changelog.yml:2]: @ycombinator 

In terms of potential release tags - do you think we should improve the spec here?

File: packages/security_detection_engine/manifest.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+categories:
+  - security
+conditions:
+  kibana.version: '^7.13.0'
+description: Rules for the detection engine in the Security application.
+format_version: 1.0.0
+license: basic
+name: security_detection_engine
+owner:
+  github: elastic/protections
```
Comment (mtojek) [inline packages/security_detection_engine/manifest.yml:10]: FYI: This team should have "write" or "maintainer" permissions to the elastic/package-storage. It's for sending notifications about waiting PRs.

**Review by mtojek — CHANGES_REQUESTED** [review]
I didn't review JSON files, as I don't have domain knowledge here and can't verify them thoroughly.

Please merge the latest master to load the latest dependency on elastic-package (it should start accepting new Kibana types).

Before merging this PR we need to make sure the PR is green.

File: packages/security_detection_engine/changelog.yml (hunk 1 above)
Comment (ycombinator) [reply packages/security_detection_engine/changelog.yml:2]: Yep, we should make the version pattern allowed in changelog [versions](https://github.com/elastic/package-spec/blob/63e8131ff323c35d756cab2cf062f094106a0de0/versions/1/changelog.spec.yml#L14) the same as the one allowed in package manifest [versions](https://github.com/elastic/package-spec/blob/63e8131ff323c35d756cab2cf062f094106a0de0/versions/1/manifest.spec.yml#L127). Even better would be to share the same pattern for both version definitions. I am making a PR to `package-spec` now to address this inconsistency.

Comment (rw-access) [reply packages/security_detection_engine/changelog.yml:2]: thanks! i'll try that out

File: packages/security_detection_engine/manifest.yml (hunk 2 above)
Comment (rw-access) [reply packages/security_detection_engine/manifest.yml:10]: thanks. what do you think is the ideal team to put in this field?
most likely, `@elastic/protections` or `@elastic/security-intelligence-analytics` will be the team thats publishing this package and the stewards of it. but if the intention of this field is something different, then i'll gladly update

File: packages/security_detection_engine/changelog.yml (hunk 1 above)
Comment (rw-access) [reply packages/security_detection_engine/changelog.yml]: updated deps in c4745655dfe856ae47278df0283cd92fa96d8300

File: go.mod [hunk 3]
```diff
@@ -3,11 +3,38 @@ module github.com/elastic/integrations
...
+	go.mongodb.org/mongo-driver v1.5.0 // indirect
+	golang.org/x/crypto v0.0.0-20210322153248-0c34fe9e7dc2 // indirect
+	golang.org/x/net v0.0.0-20210323141857-08027d57d8cf // indirect
+	golang.org/x/oauth2 v0.0.0-20210313182246-cd4f82c27b84 // indirect
+	golang.org/x/sys v0.0.0-20210320140829-1e4c9ba3b0c4 // indirect
+	golang.org/x/term v0.0.0-20210317153231-de623e64d2a6 // indirect
+	golang.org/x/text v0.3.5 // indirect
+	google.golang.org/appengine v1.6.7 // indirect
 	gopkg.in/yaml.v2 v2.4.0
+	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
```
Comment (ycombinator) [inline go.mod:39]: Could you remove all the indirect dependencies and run `go mod tidy`? It should cut down the size of `go.mod` and `go.sum` significantly.

File: packages/security_detection_engine/manifest.yml (hunk 2 above)
Comment (ycombinator) [reply packages/security_detection_engine/manifest.yml:10]: Eventually this package will need to be served up by the Elastic Package Registry. Packages that are to be served by the Elastic Package Registry are stored in the https://github.com/elastic/package-storage repository in one of its three branches (depending on their stability): `snapshot`, `staging`, or `production`.

When this PR is merged, the CI job for the `master` branch of this repo (`elastic/integrations`) will detect that this package version (`security_detection_engine 1.0.0`) does not exist in any of the three branches of the `package-storage` repository and automatically create a PR to the `snapshot` branch of the `package-storage` repository. The team that is specified in this field (`owner.github`) will get pinged on that PR so they know to review it and, if it looks good, merge it.

So it should be a team who would be best fit to review changes to this package and approve new versions of it being added to the Elastic Package Registry.

Comment (rw-access) [reply packages/security_detection_engine/manifest.yml]: Had a few conversations and figured out the most relevant owner. Updated in 3610bcd711e8ab660e8df076d2fa1c59337fdb30

File: go.mod (hunk 3 above)
Comment (rw-access) [reply go.mod]: Fixed in a8aa2286deef30932ca8bda7c61969e35ee09707

**Review by ycombinator — APPROVED** [review]
LGTM.

File: go.mod [hunk 4]
```diff
@@ -4,7 +4,7 @@ go 1.12
 
 require (
 	github.com/blang/semver v3.5.1+incompatible
-	github.com/elastic/elastic-package v0.0.0-20210323050956-96ac7e929cae
+	github.com/elastic/elastic-package v0.0.0-20210324160257-57698a40f3e3
```
Comment (ruflin) [inline go.mod:7]: Could you pull this go mod change into a speparate PR?

**Comment by mtojek** [comment]
Hm... I spotted some weird issue:

```
[2021-03-24T21:22:19.014Z] Error: checking package failed: linting package failed: found 3 validation errors:
[2021-03-24T21:22:19.014Z]    1. item [test-audit.log-config.json] is not allowed in folder [/var/lib/jenkins/workspace/gest-manager_integrations_PR-797/src/github.com/elastic/integrations/packages/gcp/data_stream/audit/_dev/test/pipeline]
[2021-03-24T21:22:19.014Z]    2. item [test-firewall.log-config.json] is not allowed in folder [/var/lib/jenkins/workspace/gest-manager_integrations_PR-797/src/github.com/elastic/integrations/packages/gcp/data_stream/firewall/_dev/test/pipeline]
[2021-03-24T21:22:19.014Z]    3. item [test-vpcflow.log-config.json] is not allowed in folder [/var/lib/jenkins/workspace/gest-manager_integrations_PR-797/src/github.com/elastic/integrations/packages/gcp/data_stream/vpcflow/_dev/test/pipeline]
```

File: go.mod (hunk 4 above)
Comment (mtojek) [reply go.mod:7]: I did it here: https://github.com/elastic/integrations/pull/804 (spotted also a different issue). Once it's merged, please rebase this PR against master.

**Comment by mtojek** [comment]
@rw-access Please rebase this branch again against the master branch. I merged the update on elastic-package, so above errors should disappear.

File: go.mod (hunk 4 above)
Comment (rw-access) [reply go.mod]: Merged in master and confirmed the go changes are gone.
Thanks @mtojek

**Comment by rw-access** [comment]
@mtojek CI is green now. thanks for pulling the deps into a separate PR.
do i have your 👍 to merge?

**Pull request merged**